x/tx/decode: (*Decoder).Decode crashes with a runtime panic slice out bounds with invalid length prefix

Found by fuzzing, the code crashes when decoding a tx whose contents have an invalid length prefix
```shell
panic: runtime error: slice bounds out of range [-1:] [recovered]
	panic: runtime error: slice bounds out of range [-1:]

goroutine 10 [running]:
testing.tRunner.func1.2({0x1957e20, 0xc000765860})
	/Users/emmanuelodeke/go/src/go.googlesource.com/go/src/testing/testing.go:1526 +0x24e
testing.tRunner.func1()
	/Users/emmanuelodeke/go/src/go.googlesource.com/go/src/testing/testing.go:1529 +0x39f
panic({0x1957e20, 0xc000765860})
	/Users/emmanuelodeke/go/src/go.googlesource.com/go/src/runtime/panic.go:884 +0x213
cosmossdk.io/x/tx/decode.RejectUnknownFields({0xc000611e50, 0xa, 0x10}, {0x1affe60, 0xc00011a740}, 0x1, {0x1ae7820, 0xc000132048})
	/Users/emmanuelodeke/go/src/github.com/cosmos/cosmos-sdk/x/tx/decode/unknown.go:93 +0x905
cosmossdk.io/x/tx/decode.(*Decoder).Decode(0xc000012bc0, {0xc000717f70, 0xc, 0x10})
	/Users/emmanuelodeke/go/src/github.com/cosmos/cosmos-sdk/x/tx/decode/decode.go:69 +0x1c6
cosmossdk.io/x/tx/decode.FuzzDecode.func1(0x0?, {0xc000717f70, 0xc, 0x10})
	/Users/emmanuelodeke/go/src/github.com/cosmos/cosmos-sdk/x/tx/decode/fuzz_test.go:121 +0xb6
reflect.Value.call({0x188ad40?, 0xc000574378?, 0x13?}, {0x19ba3a8, 0x4}, {0xc0003eff50, 0x2, 0x2?})
	/Users/emmanuelodeke/go/src/go.googlesource.com/go/src/reflect/value.go:586 +0xb07
reflect.Value.Call({0x188ad40?, 0xc000574378?, 0x1edf9b0?}, {0xc0003eff50?, 0x19b96c0?, 0xc000611e08?})
	/Users/emmanuelodeke/go/src/go.googlesource.com/go/src/reflect/value.go:370 +0xbc
testing.(*F).Fuzz.func1.1(0x0?)
	/Users/emmanuelodeke/go/src/go.googlesource.com/go/src/testing/fuzz.go:335 +0x3f3
testing.tRunner(0xc000007d40, 0xc000237c20)
	/Users/emmanuelodeke/go/src/go.googlesource.com/go/src/testing/testing.go:1576 +0x10b
created by testing.(*F).Fuzz.func1
	/Users/emmanuelodeke/go/src/go.googlesource.com/go/src/testing/fuzz.go:322 +0x5b9
exit status 2
FAIL	cosmossdk.io/x/tx/decode	0.691s
```

Fix coming up shortly.

/cc @elias-orijtech

--- a/x/tx/CHANGELOG.md
+++ b/x/tx/CHANGELOG.md
@@ -31,6 +31,10 @@ Ref: https://keepachangelog.com/en/1.0.0/
 
 ## [Unreleased]
 
+### Bug Fixes
+
+* [#16681](https://github.com/cosmos/cosmos-sdk/pull/16681): catch and fix `(*Decoder).Decode` crash from invalid length prefix in Tx bytes
+
 ## v0.8.0
 
 ### Improvements

--- a/x/tx/decode/decode_test.go
+++ b/x/tx/decode/decode_test.go
@@ -3,6 +3,7 @@ package decode_test
 import (
 	"encoding/hex"
 	"fmt"
+	"strings"
 	"testing"
 
 	"github.com/cosmos/cosmos-proto/anyutil"
@@ -117,4 +118,32 @@ func (d dummyAddressCodec) StringToBytes(text string) ([]byte, error) {
 
 func (d dummyAddressCodec) BytesToString(bz []byte) (string, error) {
 	return hex.EncodeToString(bz), nil
+}
+
+func TestDecodeTxBodyPanic(t *testing.T) {
+	crashVector := []byte{
+		0x0a, 0x0a, 0x09, 0xe7, 0xbf, 0xba, 0xe6, 0x82, 0x9a, 0xe6, 0xaa, 0x30,
+	}
+
+	cdc := new(dummyAddressCodec)
+	signingCtx, err := signing.NewContext(signing.Options{
+		AddressCodec:          cdc,
+		ValidatorAddressCodec: cdc,
+	})
+	if err != nil {
+		t.Fatal(err)
+	}
+	dec, err := decode.NewDecoder(decode.Options{
+		SigningContext: signingCtx,
+	})
+	if err != nil {
+		t.Fatal(err)
+	}
+	_, err = dec.Decode(crashVector)
+	if err == nil {
+		t.Fatal("expected a non-nil error")
+	}
+	if g, w := err.Error(), "could not consume length prefix"; !strings.Contains(g, w) {
+		t.Fatalf("error mismatch\n%s\ndoes not contain\n\t%q", g, w)
+	}
 }

--- a/x/tx/decode/unknown.go
+++ b/x/tx/decode/unknown.go
@@ -85,6 +85,15 @@ func RejectUnknownFields(bz []byte, desc protoreflect.MessageDescriptor, allowUn
 
 		// consume length prefix of nested message
 		_, o := protowire.ConsumeVarint(fieldBytes)
+		if o < 0 {
+			err = fmt.Errorf("could not consume length prefix fieldBytes for nested message: %v: %w",
+				fieldMessage, protowire.ParseError(o))
+			return hasUnknownNonCriticals, err
+		} else if o > len(fieldBytes) {
+			err = fmt.Errorf("length prefix > len(fieldBytes) for nested message: %v", fieldMessage)
+			return hasUnknownNonCriticals, err
+		}
+
 		fieldBytes = fieldBytes[o:]
 
 		var err error

